lua/neotest-haskell/cabal.lua

```diff
@@ -12,6 +12,19 @@ local function get_package_file(package_root)
   end
 end
 
+--- A 'simple' cabal project is a haskell project without project.cabal file
+--- that contains a '*.cabal' file.
+---@async
+function cabal.build_command_simple(pos)
```

Comment:
This function shares a lot of duplicate code with `cabal.build_command`.

To avoid this, we could instead reuse `cabal.build_command` and add an `opts` parameter - a `table` that has a `single_pkg_project` flag. We set it to `true` if there's no `cabal.project` file and also no `stack.yaml` file, but there is a `*.cabal` file.

Reply:
> Two lines of code is not a "a lot"

Apart from the log it's 100 % of `build_command_simple`. The table contains `cabal build` instead of `cabal new-build`, but for consistency it should be `cabal new-build` (this should be configurable, but that is another scope).

Fair point about the number of arguments. If we swap the arguments `pos` and `package_root`, we can make `package_root` optional (type `string|nil`), and there's no need for an `opts` table. We simply check if `package_root` is `nil` and not insert the package name if it is.

> Moving this logic to cabal.build_command makes that function even more complicated

I'm not proposing to move any logic to `build_command`. I'm proposing to add a conditional check so that we don't have duplicate logic and two separate call sites (which I would argue is more complex and error-prone).

At the call site (in `build_spec`), we can check if it's a stack project first, and if not do something like:

```lua
local is_multi_pkg = lib.files.exists(project_root .. '/cabal.project')
local is_simple_pkg = lib.files.match_root_pattern('*.cabal')(project_root) ~= nil
if not is_multi_pkg and not is_simple_pkg then
  logger.error('Project is neither a Cabal nor a Stack project.')
  return
end
return mkCommand(cabal.build_command(pos, is_multi_pkg and package_root))
```

Reply:
Given your second comment about the presence of a `stack.yaml` not meaning that it's a stack project, we shouldn't check if it's a stack project first.

I'll put some more thought into this tomorrow if I have some time...

Reply:
Regarding the complexity of `build_command`: I should probably extract the impure logic (apart from the logs). But that's for another PR :smile:

Reply:
> The table contains cabal build instead of cabal new-build, but for consistency it should be cabal new-build 


`cabal --help` doesn't mention the `new-` commands anymore. That was a naming scheme when the "new" way of building haskell projects was still experimental and the cabal developers didn't want to break projects by replacing the unprefixed commands. Later they added `v1-` and `v2-`-Prefixes as they are forward compatible.

If parsing the output of cabal test is is very sensitive to the cabal version, then you should probably use `cabal v2-test` instead of `cabal new-test`. Otherwise, I'd go without a prefix. 

I'll squash the functions together by making package_root optional and test it tomorrow.

Reply:
> cabal --help doesn't mention the new- commands anymore

Good to know. Thanks.